docs/redux.md

```diff
@@ -0,0 +1,1228 @@
+
+- **Priority A: Essential** - These rules help prevent errors, so learn and abide by them at all costs.
+- **Priority B: Strongly Recommended** - These rules have been found to improve readability and/or developer experience in most projects.
+- **Priority C: Recommended**
+
+## 3.1 **Priority A Rules: Essential**
+
+## 3.1.1 **Do Not Mutate State**
+
+Mutating state is the most common cause of bugs in Redux applications.
+
+```typescript
+// Bad: Mutating state directly
+const initialState = { items: [] };
+
+function myReducer(state = initialState, action) {
+  switch (action.type) {
+    case 'ADD_ITEM':
+      state.items.push(action.payload); // Direct mutation
+      return state;
+    default:
+      return state;
+  }
+}
+
+// Good: Using immutable updates
+const initialState = { items: [] };
+
+function myReducer(state = initialState, action) {
+  switch (action.type) {
+    case 'ADD_ITEM':
+      return {
+        ...state,
+        items: [...state.items, action.payload], // Immutable update
+      };
+    default:
+      return state;
+  }
+}
+```
+
+## 3.1.2 **Reducers Must Not Have Side Effects**
+
+Reducers should only depend on their state and action arguments.
+
+
+
+```typescript
+// Bad: Performing side effects in reducers
+function myReducer(state = initialState, action) {
+  switch (action.type) {
+    case 'FETCH_DATA':
+      fetch('/api/data') // Side effect
+        .then(response => response.json())
+        .then(data => {
+          state.data = data; // Direct mutation
+        });
+      return state;
+    default:
+      return state;
+  }
+}
+
+// Good: Handling side effects outside reducers
+function myReducer(state = initialState, action) {
+  switch (action.type) {
+    case 'SET_DATA':
+      return {
+        ...state,
+        data: action.payload,
+      };
+    default:
+      return state;
+  }
+}
+
+function fetchData() {
+  return (dispatch) => {
+    fetch('/api/data')
+      .then(response => response.json())
+      .then(data => {
+        dispatch({ type: 'SET_DATA', payload: data });
+      });
+  };
+}
+```
+
+## 3.1.3 **Do Not Put Non-Serializable Values in State or Actions**
+
+Avoid putting non-serializable values such as Promises, Symbols, Maps/Sets, functions, or class instances into the Redux store state or dispatched actions.
+
+```typescript
+// Bad: Storing non-serializable values in state
+const initialState = { data: new Map() };
+
+function myReducer(state = initialState, action) {
+  switch (action.type) {
+    case 'SET_DATA':
+      return {
+        ...state,
+        data: action.payload, // payload is a Map
+      };
+    default:
+      return state;
+  }
+}
+
+// Good: Storing serializable values in state
+const initialState = { data: {} };
+
+function myReducer(state = initialState, action) {
+  switch (action.type) {
+    case 'SET_DATA':
+      return {
+        ...state,
+        data: action.payload, // payload is a plain object
```

Comment:
Q: should we convert to JSON string first? Is there a preferred format?

Reply:
Converting data to a JSON string is one way to ensure data is serializable, but it is not always necessary or preferred. Instead, we should aim to use plain JavaScript objects and arrays